(#Closes 1575 and 2248) Change to line length module to avoid line continuation before any code

LonelyCat124 opened the pull request:
I think this fixes the behaviour seen in #1575 and #2248. The fix is simple, which just prevents line continuation before the first statement on a line.

I'm not clear how to have a good test for this, I'll put one together but any improvement ideas would be useful.

LonelyCat124 commented:
Ok, test added, I'm ready for review for this.

LonelyCat124 commented:
The only thing I can think of is if you had a statement that was already indented too far this wouldn't work, but there is no good solution to that problem I think.

arporter commented on src/psyclone/line_length.py:
```diff
@@ -13,9 +13,11 @@
 
 def find_break_point(line, max_index, key_list):
     ''' find the most appropriate break point for a fortran line '''
-
```
Unfortunately (for you) this is ancient code. Please could you change the copyright header to be the one we now use and document the arguments and return value/type of this method.
Please could you also remove the `from __future__` import near the top.

arporter replied:
Please could you also move the `import re` at L53 to the top (pylint complains).

LonelyCat124 replied:
Updated.
There is no `from __future__` import in my version of the file I can find.

LonelyCat124 replied:
Ah future is in the tests - i removed it there.

arporter replied:
Oops, my mistake and thanks.

arporter commented on src/psyclone/tests/line_length_test.py:
```diff
@@ -330,6 +330,32 @@ def test_edge_conditions_comments():
     assert output_string == expected_output
 
 
+def test_lone_lines_allocate(fortran_reader, fortran_writer):
```
You could keep this test but I would suggest testing the method `find_break_point` directly - you can then give it a number of different strings by e.g. parameterising the test (pytest parametrize).

LonelyCat124 replied:
Done

arporter commented on src/psyclone/tests/line_length_test.py:
```diff
@@ -330,6 +330,32 @@ def test_edge_conditions_comments():
+    '''
+
+    psyir = fortran_reader.psyir_from_source(code)
+    from psyclone.line_length import FortLineLength
```
Please move this import to the top.

LonelyCat124 replied:
Done - was already imported.

arporter reviewed (changes requested):
Thanks for fixing this @LonelyCat124. The method you've touched unfortunately needs bringing up-to-date with our current practises. Aside from that, I think it would be good (and hopefully simple) to just test that method directly for a variety of input strings.
I've triggered the integration tests so will check those next time.
This is a bug fix so no doc update is required but I'll check that the Ref Guide builds OK once you've improved the doc string of the method.

LonelyCat124 commented:
@arporter I hope I've addressed the comments correctly, pycodestyle and pylint all look good on my end now, and I believe the unit tests fully cover the functionality.

arporter commented on src/psyclone/line_length.py:
```diff
@@ -1,29 +1,78 @@
+                     for breaking the line, i.e. if the list contains multiple
+                     elements, any possible position of the first element will
+                     be found before trying any other element of the list.
+    :type key_list: list of str.
```
`List[str]` please.

LonelyCat124 replied:
Done

arporter commented on src/psyclone/line_length.py:
```diff
@@ -1,29 +1,78 @@
+    :returns: index to break the line into multiple lines.
+    :rtype: int
+
+    :raises Exception: if no suitable break point is found in line.
```
s/Exception/InternalError/

LonelyCat124 replied:
Done

arporter reviewed (changes requested):
Almost perfect now.
Just a couple of doc string fixes needed.
Ref. guide builds fine.
Integration tests are fine.
Changes are pylint clean.
Test file covers source file 100%.
Once those changes are made this can go on.

arporter reviewed (approved):
All requested changes have been made (thanks).
Will proceed to merge.
